<a href="https://colab.research.google.com/github/nikini-wanigasekara/-Statistical-Learning-e23418/blob/main/Answer_data_analysis_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Foundations of Statistical Inference & Hypothesis Testing**

Objective of this assignment is to rigorously evaluate the stationarity of an operational baseline and master the mathematical machinery required to control false-alarm risks in continuous diagnostic layers.

## **Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)**



1. Consider the sequence of i.i.d. random variables  
$$\mathbf{X}_1,\mathbf{X}_2,\cdots, \mathbf{X}_n\in \mathbb{R}^m$$
with $\boldsymbol{\mu}=\mathbb{E}(\mathbf{X}_i)$ and $\boldsymbol{\Sigma}=\mathrm{Var}(\mathbf{X}_i)$. The unobservable multivariate population parameters $(\boldsymbol{\mu}, \boldsymbol{\Sigma})$ are estimated with their corresponding unbiased empirical sample estimators $(\widehat{\boldsymbol{\mu}}_n, \mathbf{S})$, where
$$\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{X}_{i},$$
and
$$\mathbf{S} = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T .$$

* Formally derive why the raw Maximum Likelihood Estimator for the covariance matrix, $\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}$, is inherently **biased** in finite samples, by showing that:

$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n-1}{n}\boldsymbol{\Sigma}$$


* Mathematically demonstrate how applying Bessel's correction yields the unique, unbiased estimator $\mathbf{S}$.


2. Under the null hypothesis $H_0$ that a structural asset remains completely healthy (the status quo), a chosen test statistic tracks a deterministic parametric distribution.
* Define a **Type I Error ($\alpha$)** and a **Type II Error ($\beta$)** explicitly within the context of continuous structural health monitoring.
* If an engineer sets an ultra-conservative significance threshold (e.g., $\alpha = 0.0001$) to prevent false alarms from shutting down an industrial plant, explain the mathematical consequence this decision exerts on the statistical power of the diagnostic layer. How does this modify the geometric volume of your "healthy operation" confidence ellipsoid?

---
###  Answer: Part A - Q1 & Q2
---


#### **A1 — Bias of $\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}$ and Bessel's Correction**

The MLE covariance estimator is defined as:

$$\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n}\sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

**Step 1.** Decompose the centred observation using the true mean $\boldsymbol{\mu}$:
$$\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n = (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$$

**Step 2.** Define $\mathbf{d}_i = \mathbf{X}_i - \boldsymbol{\mu}$ and $\bar{\mathbf{d}} = \widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu} = \frac{1}{n}\sum_i \mathbf{d}_i$. Expanding the outer product sum:
$$\sum_{i=1}^n (\mathbf{d}_i - \bar{\mathbf{d}})(\mathbf{d}_i - \bar{\mathbf{d}})^T = \sum_{i=1}^n \mathbf{d}_i\mathbf{d}_i^T - n\bar{\mathbf{d}}\bar{\mathbf{d}}^T$$

**Step 3.** Take expectations. Since the $\mathbf{X}_i$ are i.i.d.:
$$\mathbb{E}\left[\mathbf{d}_i\mathbf{d}_i^T\right] = \boldsymbol{\Sigma}, \qquad \mathbb{E}\left[\bar{\mathbf{d}}\bar{\mathbf{d}}^T\right] = \frac{1}{n}\boldsymbol{\Sigma}$$

Therefore:
$$\mathbb{E}\left[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}\right] = \frac{1}{n}\left(n\boldsymbol{\Sigma} - n\cdot\frac{1}{n}\boldsymbol{\Sigma}\right) = \frac{n-1}{n}\boldsymbol{\Sigma} \quad \blacksquare$$

The MLE estimator systematically **underestimates** $\boldsymbol{\Sigma}$ by the factor $(n-1)/n$, confirming it is biased in finite samples.

**Bessel's Correction — deriving the unbiased estimator $\mathbf{S}$**

Rescaling by $n/(n-1)$ exactly removes the bias:
$$\mathbf{S} = \frac{n}{n-1}\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n-1}\sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

Taking expectations:
$$\mathbb{E}[\mathbf{S}] = \frac{n}{n-1}\cdot\frac{n-1}{n}\boldsymbol{\Sigma} = \boldsymbol{\Sigma} \quad \blacksquare$$

The denominator $(n-1)$ replaces $n$ because estimating $\widehat{\boldsymbol{\mu}}_n$ from the same data consumes one degree of freedom.

---

#### **A2 — Type I / II Errors and the Effect of Ultra-Conservative $\alpha$**

**Type I Error ($\alpha$):** Rejecting $H_0$ when the asset is actually healthy — a false alarm that triggers an unnecessary shutdown.

**Type II Error ($\beta$):** Failing to reject $H_0$ when the structure has genuinely degraded — a missed detection allowing a real fault to go unnoticed.

**Mathematical consequence of setting $\alpha = 0.0001$:**

The decision boundary is set at the $(1-\alpha)$ quantile of the reference distribution:
$$c_\alpha = F^{-1}(1-\alpha)$$

Lowering $\alpha$ pushes $c_\alpha$ to a much larger value. Since power and $\alpha$ are inversely linked through the non-centrality parameter at fixed $n$:
$$\text{Power} = 1-\beta = P(\text{reject }H_0 \mid H_1 \text{ true})$$

reducing $\alpha$ while holding $n$ fixed forces $\beta$ to increase — smaller anomalies that were previously detectable are now missed.

**Geometric interpretation — the confidence ellipsoid:**

The healthy-operation region is:
$$\mathcal{E}_\alpha = \left\{\mathbf{x} : (\mathbf{x}-\widehat{\boldsymbol{\mu}}_n)^T \mathbf{S}^{-1}(\mathbf{x}-\widehat{\boldsymbol{\mu}}_n) \leq c_\alpha \right\}$$

As $\alpha \to 0$, the critical radius $\sqrt{c_\alpha}$ grows. The ellipsoid **expands**, enclosing a larger volume of the observation space. More anomalous readings are incorrectly classified as healthy — the false-alarm rate is reduced at the direct cost of masking real structural faults.


## **Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)**

The Lindeberg-Lévy Multivariate Central Limit Theorem establishes that as the sample snapshot size approaches infinity ($n \to \infty$), the standardized difference between the empirical mean and population mean converges in distribution to a Multivariate Gaussian:

$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$



However, the true population covariance matrix $\boldsymbol{\Sigma}$ is a latent, unobservable parameter.
* State **Slutsky’s Theorem** rigorously.
* Prove how Slutsky's Theorem justifies substituting the empirical sample covariance matrix $\mathbf{S}$ for the latent population matrix $\boldsymbol{\Sigma}$ as $n \to \infty$, thereby operationalizing the practical parametric modeling distribution:

$$\widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left({\boldsymbol{\mu}}, \frac{1}{n}\mathbf{S}\right)$$







---
###  Answer: Part B
---


**Slutsky's Theorem (rigorous statement):**

Let $\{X_n\}$ and $\{Y_n\}$ be sequences of random vectors/matrices. If:
$$X_n \xrightarrow{d} X, \qquad Y_n \xrightarrow{p} c \text{ (a constant)}$$
then:
- $X_n + Y_n \xrightarrow{d} X + c$
- $Y_n X_n \xrightarrow{d} cX$
- $Y_n^{-1} X_n \xrightarrow{d} c^{-1}X$ (provided $c$ is invertible)

**Applying Slutsky's Theorem to substitute $\mathbf{S}$ for $\boldsymbol{\Sigma}$:**

By the Lindeberg-Lévy Multivariate CLT:
$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathcal{N}(\mathbf{0}, \boldsymbol{\Sigma})$$

By the Law of Large Numbers, the sample covariance converges in probability to the true covariance:
$$\mathbf{S} \xrightarrow{p} \boldsymbol{\Sigma}$$

Since $\mathbf{S}$ converges in probability to a constant matrix $\boldsymbol{\Sigma}$, Slutsky's Theorem permits substituting $\mathbf{S}$ for $\boldsymbol{\Sigma}$ in the limiting distribution without changing the limit. Specifically:
$$\sqrt{n}\,\mathbf{S}^{-1/2}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathcal{N}(\mathbf{0}, \mathbf{I})$$

This operationalises the practical working distribution:
$$\widehat{\boldsymbol{\mu}}_n \overset{\text{approx}}{\sim} \mathcal{N}\!\left(\boldsymbol{\mu},\, \frac{1}{n}\mathbf{S}\right) \quad \blacksquare$$

valid for sufficiently large $n$ even though $\boldsymbol{\Sigma}$ is unknown and replaced by its consistent estimator $\mathbf{S}$.


## **Part C: Numerical Verification**

***Scenario:** You are provided an industrial dataset containing $n = 5,000$ snapshots across $m = 3$ sensor channels measuring physical strain. To verify the weak i.i.d. assumption before initializing a global PCA or Factor Analysis layer, you must explicitly test **Condition 1: Constant Joint Mean (First Moment Homogeneity)** across the time index.*

**Your Task:**
Write a clean, self-contained Python function within Google Colab to partition the dataset into $g = 5$ consecutive, non-overlapping blocks (chunks) of size $n_j = 1,000$. Implement a Multivariate Analysis of Variance (MANOVA) tracking framework from scratch (or leveraging `statsmodels`) to calculate **Wilks' Lambda ($\Lambda$)**.

$$\Lambda = \frac{|\mathbf{W}|}{|\mathbf{B} + \mathbf{W}|}$$

Transform your calculated $\Lambda$ using **Bartlett’s $\chi^2$ approximation**:


$$\chi^2_{\text{calc}} = -\left(n - 1 - \frac{m + g}{2}\right) \ln \Lambda$$

Execute your code using the synthetic data generator provided below. Report your exact calculated Wilks' $\Lambda$, the Bartlett $\chi^2$ test statistic, and its corresponding $p$-value under $\chi^2 \sim \chi^2_{m(g-1)}$. Based on an alpha threshold of $\alpha = 0.05$, state your final diagnostic conclusion: Has the baseline center shifted over time, or is the first moment homogeneous?



```python
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

# Generating base stationary multivariate normal noise
base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

# Injecting a subtle, hidden structural drift in the final 1,000 snapshots
base_data[4000:, 0] += 0.015  # Sensor 1 drift
base_data[4000:, 2] -= 0.010  # Sensor 3 drift

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partitions the dataset into g chunks and evaluates first-moment homogeneity
    via Wilks' Lambda and Bartlett's Chi-Square asymptotic transformation.
    """
    # YOUR CODE GOES HERE
    # 1. Split df into g equal consecutive blocks
    # 2. Compute the global mean and individual chunk means
    # 3. Calculate Within-Chunk Matrix (W) and Between-Chunk Matrix (B)
    # 4. Compute Wilks' Lambda, Bartlett's Chi-Square, and the final p-value
    pass

# Execute your completed function below:
# results = verify_first_moment_homogeneity(df_strain, g_chunks=5)
# print(results)
```

---
###  Answer: Part C
---


In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)
base_data[4000:, 0] += 0.015
base_data[4000:, 2] -= 0.010
df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partitions the dataset into g chunks and evaluates first-moment homogeneity
    via Wilks' Lambda and Bartlett's Chi-Square asymptotic transformation.
    """
    n, m = df.shape
    n_j = n // g_chunks
    chunks = [df.iloc[i * n_j:(i + 1) * n_j].values for i in range(g_chunks)]

    grand_mean = df.values.mean(axis=0)

    # Within-group scatter matrix W
    W = np.zeros((m, m))
    for chunk in chunks:
        chunk_mean = chunk.mean(axis=0)
        centered = chunk - chunk_mean
        W += centered.T @ centered

    # Between-group scatter matrix B
    B = np.zeros((m, m))
    for chunk in chunks:
        chunk_mean = chunk.mean(axis=0)
        diff = (chunk_mean - grand_mean).reshape(-1, 1)
        B += n_j * (diff @ diff.T)

    # Wilks' Lambda
    wilks_lambda = np.linalg.det(W) / np.linalg.det(B + W)

    # Bartlett's chi-square approximation
    dof = m * (g_chunks - 1)
    chi2_calc = -(n - 1 - (m + g_chunks) / 2) * np.log(wilks_lambda)
    p_value = 1 - stats.chi2.cdf(chi2_calc, df=dof)

    conclusion = (
        "REJECT H0: Baseline centre has shifted — first moment is NOT homogeneous."
        if p_value < 0.05 else
        "FAIL TO REJECT H0: First moment is homogeneous."
    )

    return {
        "wilks_lambda":       round(wilks_lambda, 6),
        "chi2_stat":          round(chi2_calc, 4),
        "degrees_of_freedom": dof,
        "p_value":            p_value,
        "conclusion":         conclusion
    }

results = verify_first_moment_homogeneity(df_strain, g_chunks=5)
for k, v in results.items():
    print(f"{k:25s}: {v}")


wilks_lambda             : 0.990452
chi2_stat                : 47.9215
degrees_of_freedom       : 12
p_value                  : 3.2255259508895406e-06
conclusion               : REJECT H0: Baseline centre has shifted — first moment is NOT homogeneous.


**Reported Results:**

| Metric | Value |
|--------|-------|
| Wilks' $\Lambda$ | **0.9905** |
| Bartlett $\chi^2$ | **47.92** |
| Degrees of freedom | **12** (= $m \times (g-1) = 3 \times 4$) |
| $p$-value | **3.23 × 10⁻⁶** |

**Diagnostic conclusion ($\alpha = 0.05$):**  
Since $p = 3.23 \times 10^{-6} \ll 0.05$, we **reject $H_0$**. The injected drift in Block 5 (Sensor 1: +0.015, Sensor 3: −0.010) is successfully detected. The baseline centre has shifted; **first-moment homogeneity does not hold**.


# **Geometric Subspace Optimization via Principal Component Analysis (PCA)**

The objective of this assignment is to master the rigorous linear algebra mechanics underpinning coordinate transformations, verify total variance conservation laws, and design operational monitoring thresholds using the Hotelling's $T^2$ and $Q$ (SPE) statistics.

## **Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)**

1. Let $\widetilde{\mathbf{X}}_i = (\mathbf{X}_i - \boldsymbol{\mu}) \in \mathbb{R}^m$ represent a zero-mean, centered baseline feature vector with a symmetric, positive-semidefinite population covariance matrix $\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$. We apply the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$, and define the orthogonal projection mapping:

$$\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i$$


* Prove step-by-step that the theoretical covariance matrix of the transformed vector, $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$, simplifies exactly to the strictly diagonal matrix $\mathbf{\Lambda}$.
* Explain the precise geometric and statistical meaning of this diagonal result regarding the cross-covariances between distinct coordinates $Z_{i,j}$ and $Z_{i,k}$ (for $j \neq k$).


2. The total variance of a multivariate system can be measured in its raw coordinate space or its transformed principal component space.
* Utilizing the algebraic properties of the matrix trace ($\text{tr}$), specifically its invariance under cyclic permutations, formally prove that the total system variance is preserved during the mapping from raw space to the de-correlated principal space:

$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j$$


* If we partition the empirical matrix into an explained subspace ($\widehat{\mathbf{P}}_k \in \mathbb{R}^{m \times k}$) and an unexplained residual subspace ($\widehat{\mathbf{P}}_{m-k} \in \mathbb{R}^{m \times (m-k)}$), state the rigorous mathematical formulations for the **Cumulative Explained Variance Ratio $\Phi(k)$** and the **Residual Unexplained Variance Ratio $\Psi(k)$**.


3. In structural health monitoring (SHM), we truncate our coordinate system to retain only the first $k$ components ($k < m$). The centered raw vector is partitioned as $(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) = \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$.
* Define the reconstructed vector $\widehat{\mathbf{x}}_i$ and the residual error vector $\mathbf{e}_i$.
* Prove via the Pythagorean identity that the total squared Euclidean norm of the centered observation vector cleanly decomposes into independent subspace energies:

$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$


* Contrast the diagnostic functions of Hotelling's $T^2$ statistic and the residual $Q$ statistic (SPE). If a structure encounters an environmental load anomaly (e.g., extreme operational wind loads) versus an internal structural fracture (e.g., a fatigue crack changing the structural load pathways), match each event type to the statistic ($T^2$ or $Q$) most sensitive to it and justify your choices.

---
###  Answer: Part A - Q1, Q2, Q3
---


#### **A1 — Covariance of the Transformed Vector $\mathbf{Z}_i$**

**Claim:** $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \boldsymbol{\Lambda}$

**Proof:**

Since $\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i$ and $\mathbf{P}$ is orthogonal ($\mathbf{P}^T\mathbf{P} = \mathbf{I}$):

$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{P}^T\,\mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]\,\mathbf{P} = \mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P}$$

Substituting $\boldsymbol{\Sigma} = \mathbf{P}\boldsymbol{\Lambda}\mathbf{P}^T$:
$$= \mathbf{P}^T(\mathbf{P}\boldsymbol{\Lambda}\mathbf{P}^T)\mathbf{P} = (\mathbf{P}^T\mathbf{P})\boldsymbol{\Lambda}(\mathbf{P}^T\mathbf{P}) = \mathbf{I}\boldsymbol{\Lambda}\mathbf{I} = \boldsymbol{\Lambda} \quad \blacksquare$$

**Geometric and statistical meaning:**  
$\boldsymbol{\Lambda}$ is strictly diagonal, so every off-diagonal entry $\mathbb{E}[Z_{i,j}Z_{i,k}] = 0$ for $j \neq k$. The principal axes (columns of $\mathbf{P}$) are mutually orthogonal directions of pure variance. After projection, each coordinate $Z_{i,j}$ carries exactly $\lambda_j$ of variance and is **uncorrelated** with every other principal coordinate, transforming a correlated system into a set of statistically independent scalar channels (under Gaussianity).

---

#### **A2 — Trace Invariance and Variance Conservation**

**Proof that $\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\boldsymbol{\Lambda})$:**

Using the cyclic permutation property of the trace, $\text{tr}(\mathbf{AB}) = \text{tr}(\mathbf{BA})$:
$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{P}\boldsymbol{\Lambda}\mathbf{P}^T) = \text{tr}(\boldsymbol{\Lambda}\mathbf{P}^T\mathbf{P}) = \text{tr}(\boldsymbol{\Lambda}) = \sum_{j=1}^m \lambda_j \quad \blacksquare$$

Total variance is preserved exactly under the orthogonal PCA mapping.

**Cumulative Explained Variance Ratio $\Phi(k)$ and Residual Unexplained Variance Ratio $\Psi(k)$:**

$$\Phi(k) = \frac{\sum_{j=1}^k \widehat{\lambda}_j}{\sum_{j=1}^m \widehat{\lambda}_j}, \qquad \Psi(k) = 1 - \Phi(k) = \frac{\sum_{j=k+1}^m \widehat{\lambda}_j}{\sum_{j=1}^m \widehat{\lambda}_j}$$

---

#### **A3 — Pythagorean Decomposition, $T^2$ and $Q$**

**Reconstructed vector and residual:**
$$\widehat{\mathbf{x}}_i = \widehat{\boldsymbol{\mu}}_n + \widehat{\mathbf{P}}_k \mathbf{z}_{i,k}, \qquad \mathbf{e}_i = (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) - \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} = \widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}$$

**Pythagorean decomposition proof:**

Since $\widehat{\mathbf{P}}_k$ and $\widehat{\mathbf{P}}_{m-k}$ span orthogonal subspaces ($\widehat{\mathbf{P}}_k^T\widehat{\mathbf{P}}_{m-k} = \mathbf{0}$), the cross-term vanishes. Using orthonormality of $\mathbf{P}$:
$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\widehat{\mathbf{P}}_k\mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2 \quad \blacksquare$$

**Diagnostic roles:**

| Statistic | Formula | Monitors |
|-----------|---------|----------|
| Hotelling's $T^2$ | $T^2_i = \sum_{j=1}^k z_{i,j}^2/\widehat{\lambda}_j$ | Mahalanobis distance **within** the retained subspace |
| $Q$ (SPE) | $Q_i = \sum_{j=k+1}^m z_{i,j}^2$ | Residual energy in the **unexplained** subspace |

**Event matching:**
- **Extreme wind load (environmental anomaly):** Total load magnitude increases along the dominant structural modes already captured by the PCA subspace → **$T^2$ is the more sensitive statistic.**
- **Fatigue crack (internal fracture):** Crack redirects load between elements, creating new covariance patterns outside the nominal subspace; residual vector $\mathbf{e}_i$ grows large → **$Q$ (SPE) is the more sensitive statistic.**


## **Part B: Numerical Verification**



***Scenario:** You are building a real-time diagnostics layer for a structural component monitored by $m = 4$ sensors. You have a baseline archive of $n = 3,000$ nominal snapshots. To establish operational boundaries, you must write an optimization pipeline that analyzes how Hotelling’s $T^2$ and the residual $Q$ statistic respond to different truncation boundaries $k \in \{1, 2, 3\}$.*

**Your Task:**
Complete the Python class method below inside Google Colab. The function must:

1. Standardize and center the incoming realizations.
2. Decompose the unbiased sample covariance matrix $\mathbf{S}$ using `np.linalg.eigh` and sort components in descending order ($\widehat{\lambda}_1 \ge \dots \ge \widehat{\lambda}_m$).
3. Loop through every possible subspace cutoff value $k \in \{1, 2, 3\}$, calculating the sample vector array for Hotelling’s $T^2$ ($T^2_i = \sum_{j=1}^k \frac{z_{i,j}^2}{\widehat{\lambda}_j}$) and the $Q$ statistic ($Q_i = \sum_{j=k+1}^m z_{i,j}^2$).
4. Compute the expected value (empirical mean) of $T^2$ and $Q$ across all samples for each $k$.
5. Return these metrics to verify your theoretical derivations in Part A.



---
###  Answer: Part B
---


In [2]:
import numpy as np
import pandas as pd

def pca_diagnostic_metrics(df: pd.DataFrame, k_values=(1, 2, 3)) -> dict:
    """
    Standardizes data, decomposes the unbiased sample covariance via eigh,
    and computes Mean T² and Mean Q for each truncation cutoff k.
    """
    X = df.values.astype(float)
    mu = X.mean(axis=0)
    sigma = X.std(axis=0, ddof=1)
    Z = (X - mu) / sigma                       # standardize

    S = np.cov(Z, rowvar=False)                # unbiased covariance

    eigenvalues, eigenvectors = np.linalg.eigh(S)
    idx = np.argsort(eigenvalues)[::-1]         # sort descending
    eigenvalues  = eigenvalues[idx]
    eigenvectors = eigenvectors[:, idx]

    all_scores = Z @ eigenvectors               # (n, m) full score matrix
    total_var  = eigenvalues.sum()

    results = {}
    for k in k_values:
        T2 = np.sum((all_scores[:, :k] ** 2) / eigenvalues[:k], axis=1)
        Q  = np.sum(all_scores[:, k:] ** 2, axis=1)
        results[k] = {
            'mean_T2':       round(float(T2.mean()), 4),
            'mean_Q':        round(float(Q.mean()),  4),
            'cum_var_pct':   round(float(eigenvalues[:k].sum() / total_var * 100), 2),
            'eigenvalues':   np.round(eigenvalues, 6).tolist()
        }
    return results

# --- Synthetic 4-sensor, 3000-sample baseline archive ---
np.random.seed(42)
n, m = 3000, 4
raw = np.random.multivariate_normal(
    mean=np.zeros(m),
    cov=[[1.0, 0.8, 0.5, 0.1],
         [0.8, 1.0, 0.4, 0.1],
         [0.5, 0.4, 1.0, 0.2],
         [0.1, 0.1, 0.2, 1.0]],
    size=n
)
df_baseline = pd.DataFrame(raw, columns=[f'Sensor_{i+1}' for i in range(m)])

metrics = pca_diagnostic_metrics(df_baseline)
print(f"{'k':>3}  {'Mean T²':>10}  {'Mean Q':>10}  {'Cumul. Var %':>14}")
print("-" * 42)
for k, v in metrics.items():
    print(f"{k:>3}  {v['mean_T2']:>10.4f}  {v['mean_Q']:>10.4f}  {v['cum_var_pct']:>14.2f}")


  k     Mean T²      Mean Q    Cumul. Var %
------------------------------------------
  1      0.9997      1.8105           54.72
  2      1.9993      0.8211           79.47
  3      2.9990      0.1905           95.24


# **Latent Subspace Decomposition via Factor Analysis (FA)**

The objective is to master the generative model equations of Factor Analysis, derive factor scores from scratch using conditional expectations (Thomson's method), and numerically isolate sensor malfunctions from true structural damage by tracking communalities ($h^2$) and uniquenesses ($\varphi^2$).

## **Part A: Theoretical Fundamentals (Generative Model & Commonalities)**

1. Consider your structured data setup where $\mathbf{Z}_i \in \mathbb{R}^m$ is a normalized, zero-mean baseline feature vector whose population covariance matrix represents the universal correlation matrix $\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$. The Factor Analysis model posits the linear generative structure:
$$\mathbf{Z}_i = \boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i$$
where $\mathbf{F}_i \in \mathbb{R}^k$ are the orthogonal common factors ($\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] = \mathbf{I}_{k \times k}$) and $\boldsymbol{\epsilon}_i \in \mathbb{R}^m$ is the diagonalized measurement noise matrix ($\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi} = \text{diag}(\varphi_1^2, \dots, \varphi_m^2)$).
* Utilizing the constraint that the latent factors and errors are independent ($\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}$), prove the **Fundamental Equation of Factor Analysis**:

$$\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$


* Focus on the $j$-th diagonal entry of this decomposition. Explicitly define the **Communality ($h_j^2$)** and **Uniqueness ($\varphi_j^2$)** parameters, and explain their physical interpretations in an asset diagnostics system.


2. In Factor Analysis, we often apply an orthogonal coordinate rotation to the factor loading matrix (such as the **Varimax** rotation criterion) to improve model interpretability.
* Contrast how a raw principal component loading vector differs conceptually from a Varimax-rotated factor loading vector.
* Explain why orthogonal rotations can alter the individual columns of the loading matrix $\boldsymbol{\Lambda}$ without changing the calculated communality ($h_j^2$) of any sensor or modifying the global covariance approximation $\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$.



1. Because the common factors $\mathbf{F}_i$ are unobservable latent parameters, they cannot be directly extracted via geometric projections like PCA. Instead, we must treat them as hidden random vectors and calculate their realization using **Thomson’s Regression Method**, which relies on the conditional expectation $\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i | \mathbf{z}_i]$.
* Construct the joint partitioned vector $\mathbf{Y}_i = [\mathbf{Z}_i^T, \mathbf{F}_i^T]^T$ and prove that its joint population covariance matrix expands exactly to:

$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix}$$


* Under the standard assumption of a Multivariate Normal distribution, the conditional mean of a partitioned system is governed by the projection identity $\mathbb{E}[\mathbf{X}_2 | \mathbf{x}_1] = \boldsymbol{\mu}_2 + \boldsymbol{\Sigma}_{21} \boldsymbol{\Sigma}_{11}^{-1} (\mathbf{x}_1 - \boldsymbol{\mu}_1)$. Using this theorem, derive the final score matrix equation for Thomson’s estimator:

$$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{z}_i$$


---
### Answer: Part A - Q1, Q2, Q3
---


#### **A1 — Fundamental Equation of Factor Analysis**

**Claim:** $\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$

**Proof:**
$$\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)^T]$$
$$= \boldsymbol{\Lambda}\underbrace{\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T]}_{=\mathbf{I}}\boldsymbol{\Lambda}^T + \boldsymbol{\Lambda}\underbrace{\mathbb{E}[\mathbf{F}_i\boldsymbol{\epsilon}_i^T]}_{=\mathbf{0}} + \underbrace{\mathbb{E}[\boldsymbol{\epsilon}_i\mathbf{F}_i^T]}_{=\mathbf{0}}\boldsymbol{\Lambda}^T + \underbrace{\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T]}_{=\boldsymbol{\Psi}} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi} \quad \blacksquare$$

**Communality and Uniqueness (j-th diagonal entry):**
$$1 = \underbrace{\sum_{r=1}^k \lambda_{jr}^2}_{h_j^2} + \underbrace{\varphi_j^2}_{\text{uniqueness}}$$

- **Communality $h_j^2$:** proportion of sensor $j$'s variance explained by the $k$ shared latent factors. High $h_j^2$ means the sensor is driven by the underlying structural physics.
- **Uniqueness $\varphi_j^2 = 1 - h_j^2$:** sensor-specific noise floor. $\varphi_j^2 \approx 1$ flags an electrical fault or a sensor that does not respond to the monitored structural modes.

---

#### **A2 — Varimax Rotation vs. PCA Loadings**

A raw PCA loading vector (eigenvector) is constrained to the direction of maximum residual variance, producing dense vectors where every sensor has a non-trivial loading on every component. Varimax applies an orthogonal rotation matrix $\mathbf{T}$ to the initial loading estimate, maximising:
$$V = \frac{1}{m}\sum_{r=1}^k\left[\sum_j \lambda_{jr}^4 - \frac{1}{m}\left(\sum_j \lambda_{jr}^2\right)^2\right]$$

This drives each rotated column toward "simple structure": large loadings on a small cluster of sensors and near-zero loadings on the rest.

**Why rotation does not change $h_j^2$ or the covariance approximation:**

An orthogonal rotation replaces $\boldsymbol{\Lambda}$ with $\boldsymbol{\Lambda}^* = \boldsymbol{\Lambda}\mathbf{T}$ where $\mathbf{T}^T\mathbf{T} = \mathbf{I}$:
$$\boldsymbol{\Lambda}^*{\boldsymbol{\Lambda}^*}^T = \boldsymbol{\Lambda}\mathbf{T}\mathbf{T}^T\boldsymbol{\Lambda}^T = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$$

The product $\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$ is invariant to orthogonal rotation, so $\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$ is unchanged, and $h_j^2$ (the $j$-th diagonal of $\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$) is also rotation-invariant. Rotation only redistributes the shared variance **among** factors; it does not alter how much is shared in total.

---

#### **A3 — Thomson's Regression Derivation**

**Joint covariance of $\mathbf{Y}_i = [\mathbf{Z}_i^T, \mathbf{F}_i^T]^T$:**

- $\text{Cov}(\mathbf{Z}_i, \mathbf{Z}_i^T) = \mathbf{R}$
- $\text{Cov}(\mathbf{Z}_i, \mathbf{F}_i^T) = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)\mathbf{F}_i^T] = \boldsymbol{\Lambda}\mathbf{I} + \mathbf{0} = \boldsymbol{\Lambda}$
- $\text{Cov}(\mathbf{F}_i, \mathbf{F}_i^T) = \mathbf{I}_{k \times k}$

$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix} \quad \blacksquare$$

**Deriving Thomson's estimator:**

Under Multivariate Normality, with $\boldsymbol{\mu}_F = \mathbf{0}$, $\boldsymbol{\mu}_Z = \mathbf{0}$, $\boldsymbol{\Sigma}_{21} = \boldsymbol{\Lambda}^T$, $\boldsymbol{\Sigma}_{11} = \mathbf{R}$:
$$\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i \mid \mathbf{z}_i] = \mathbf{0} + \boldsymbol{\Lambda}^T\mathbf{R}^{-1}(\mathbf{z}_i - \mathbf{0}) = \boldsymbol{\Lambda}^T\mathbf{R}^{-1}\mathbf{z}_i \quad \blacksquare$$


## **Part B: Parametric Factor Partition Engine & UI Architecture**

The objective is to implement an end-to-end data processing and diagnostic visualization pipeline that instantiates a parametric **Factor Analysis (FA)** engine with **Varimax rotation**. The engine must partition empirical observation metrics into a shared structural subspace and a distinct diagonal noise floor, mapping the results onto a highly structured, asymmetric grid interface.

### Develop a Mathematical Engine & Pre-processing Guardrails


Implement an isolated algorithm to structure and decompose empirical data structures according to the following guidelines:

1. **Z-Score Mapping Layer:** Accept an arbitrary data frame and isolate numeric observation columns. Standardize the observations into standard normal space $Z \in \mathbb{R}^{n \times m}$ applying an unbiased sample standard deviation ($ddof=1$). Build an explicit numerical guardrail that catches zero-variance channels ($\sigma_j = 0$) and caps them to a minimum boundary threshold of $10^{-15}$ to prevent divide-by-zero runtime exceptions.
2. **Dimensionality Constraint Layer:** Let $m$ represent the count of isolated sensor features, and $k$ represent the target count of latent physical factors. Program two validation rule exceptions:
* Reject processing if the operational snapshot count ($n$) is less than or equal to the channel count ($m$).
* Throw an descriptive error if the requested latent space dimension ($k$) is not strictly less than the available feature space dimension ($m$).


3. **Information Decomposition Layer:** Fit a Factor Analysis framework utilizing a Maximum Likelihood or Singular Value Decomposition (SVD) algorithm combined with an orthogonal Varimax rotation matrix ($T$). From the fitted parameters, compute the vector arrays for:
* **Communality Vector ($h^2 \in \mathbb{R}^m$):** The total shared variance captured across each channel.
* **Uniqueness Vector ($\varphi^2 \in \mathbb{R}^m$):** The remaining isolated sensor noise floor.

### Latent Score Projection

1. **Mathematical State Reconstruction:** Using Thomson’s Minimum Mean Squared Error (MMSE) regression methodology, transform the standard normal snapshot array ($Z$) directly into hidden factor states $F \in \mathbb{R}^{n \times k}$ using the system relation:

$$F = Z R^{-1} \Lambda_{\text{rotated}}$$


2. **Performance Diagnostic Telemetry:** Calculate the empirical variance of the reconstructed factor scores enforcing an unbiased scale correction ($ddof=1$). Print a clean confirmation breakdown out to the logging console that presents the calculated system-wide metrics:

$$\text{Average System Communality \%} = \frac{1}{m}\sum_{j=1}^m h_j^2 \times 100$$


$$\text{Average System Uniqueness \%} = \frac{1}{m}\sum_{j=1}^m \varphi_j^2 \times 100$$

### Layout Engineering & Interface Design

Render a unified, production-grade **2 × 2 diagnostic canvas** using Plotly Subplots that enforces the following strict geometric layout constraints to ensure maximum readability and prevent overlapping label blocks:

* **Grid Specifications:** Configure the dashboard matrix with an explicit `horizontal_spacing=0.24` and a `vertical_spacing=0.28`. Set the global canvas parameters to `width=1250` and `height=750` utilizing a clean `plotly_white` theme.
* **Legend Formatting:** Place a single, horizontally oriented legend centered directly above the subplots (`orientation="h", x=0.5, y=1.02`). Enforce clean `margin` padding structures to isolate the interface elements (`t=150, b=60, l=140, r=80`).

Populate the grid layout using the four specific tracking profiles below:

```
                  2 × 2 FACTOR DIAGNOSTICS GRID MATRIX
+------------------------------------------+------------------------------------------+
|             [Row 1, Col 1]               |             [Row 1, Col 2]               |
|      Structural Loadings Heatmap         |     Horizontal Stacked Bar Chart         |
|   - Absolute values |λ_(j,r)|            |   - Shared Communality (h²) in Blue      |
|   - Color scale: 'YlOrRd'                |   - Isolated Uniqueness (φ²) in Orange   |
|   - Left-anchored scale bar (x = -0.15)  |   - Total space constraints: 0 to 100%   |
+------------------------------------------+------------------------------------------+
|             [Row 2, Col 1]               |             [Row 2, Col 2]               |
|         Sensor Noise Floor Line          |       Latent Empirical Variance          |
|   - Dot-dashed red line graph with markers|   - Vertical distribution bar chart      |
|   - Tracks exact Uniqueness values (φ²)  |   - Verifies energy level allocations    |
|   - Enforce tick angle adjustments (25°) |   - Black boundary outlines (width = 0.5)|
+------------------------------------------+------------------------------------------+

```

1. **Subplot (1,1): Structural Loadings Matrix Heatmap**
Render the absolute value of the rotated component weights matrix ($|\Lambda_{\text{rotated}}|$) using a continuous `YlOrRd` (Yellow-Orange-Red) color palette. To protect the clarity of the sensor text identifiers along the y-axis, isolate the continuous scale colorbar to the left side of the canvas by explicitly setting `x=-0.15`, `len=0.38`, `y=0.78`, `yanchor="middle"`, and `xanchor="right"`.
2. **Subplot (1,2): Variance Partitioning Profile**
Build a horizontal stacked bar chart mapping out the tracking allocation per sensor channel. The base layer must represent the percentage contribution of shared communalities ($h^2 \times 100$) utilizing color hex code `#1f77b4`. Stack the corresponding sensor uniqueness values ($\varphi^2 \times 100$) directly on top utilizing color hex code `#ff7f0e`. Force the x-axis limits strictly between `[0, 100]`.
3. **Subplot (2,1): Sensor Uniqueness Line Profile**
Construct a clear line chart tracking the isolated channel noise thresholds ($\varphi^2$) across all monitored hardware positions. Apply a red (`#d62728`), 2-pixel wide dot-dashed line string combined with visual `x` indicators. Apply a custom horizontal label layout optimization by rotating the category ticks to a distinct `tickangle=25` constraint.
4. **Subplot (2,2): Latent Factor Variance Distribution**
Draw a vertical bar chart displaying the empirical variance computed across the extracted common factor score vectors. Color the bars green (`#2ca02c`) with an explicit black border envelope (`width=0.5`).

---
### Answer: Part B
---


In [3]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
warnings.filterwarnings('ignore')

# ─────────────────────────────────────────────
# 1. Z-Score Mapping Layer
# ─────────────────────────────────────────────
def zscore_layer(df: pd.DataFrame):
    numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    X = df[numeric_cols].values.astype(float)
    means = X.mean(axis=0)
    stds  = X.std(axis=0, ddof=1)
    stds  = np.where(stds == 0, 1e-15, stds)   # zero-variance guardrail
    Z = (X - means) / stds
    return Z, numeric_cols

# ─────────────────────────────────────────────
# 2. Dimensionality Constraint Layer
# ─────────────────────────────────────────────
def check_dimensions(n: int, m: int, k: int):
    if n <= m:
        raise ValueError(
            f"Snapshot count n={n} must exceed feature count m={m}. "
            "Provide more observations."
        )
    if k >= m:
        raise ValueError(
            f"Latent dimension k={k} must be strictly less than feature dimension m={m}."
        )

# ─────────────────────────────────────────────
# 3. Information Decomposition Layer
# ─────────────────────────────────────────────
def fa_decomposition(Z: np.ndarray, k: int):
    fa = FactorAnalysis(n_components=k, rotation='varimax', random_state=42)
    fa.fit(Z)
    loadings      = fa.components_.T          # shape (m, k)
    communalities = np.sum(loadings ** 2, axis=1)
    uniqueness    = 1 - communalities
    return loadings, communalities, uniqueness

# ─────────────────────────────────────────────
# 4. Thomson's MMSE Score Projection
# ─────────────────────────────────────────────
def thomson_scores(Z: np.ndarray, loadings: np.ndarray) -> np.ndarray:
    R = np.corrcoef(Z, rowvar=False)
    F = Z @ np.linalg.inv(R) @ loadings   # F = Z R⁻¹ Λ_rotated
    return F

# ─────────────────────────────────────────────
# 5. Main Engine
# ─────────────────────────────────────────────
def run_fa_engine(df: pd.DataFrame, k: int = 2):
    Z, sensor_names = zscore_layer(df)
    n, m = Z.shape
    check_dimensions(n, m, k)

    loadings, h2, phi2 = fa_decomposition(Z, k)
    F                  = thomson_scores(Z, loadings)
    var_factors        = F.var(axis=0, ddof=1)

    print("=" * 55)
    print(f"  Average System Communality : {h2.mean()*100:.2f}%")
    print(f"  Average System Uniqueness  : {(1-h2).mean()*100:.2f}%")
    print(f"  {'Sensor':15s}  {'h²':>8}  {'φ²':>8}")
    for j, name in enumerate(sensor_names):
        print(f"  {name:15s}  {h2[j]:>8.4f}  {phi2[j]:>8.4f}")
    print(f"\n  Factor Score Variances: {np.round(var_factors, 4)}")
    print("=" * 55)

    return loadings, h2, phi2, F, var_factors, sensor_names

# ─────────────────────────────────────────────
# 6. 2×2 Diagnostic Dashboard
# ─────────────────────────────────────────────
def plot_fa_dashboard(loadings, h2, phi2, var_factors, sensor_names):
    factor_names = [f"Factor {r+1}" for r in range(loadings.shape[1])]

    fig = make_subplots(
        rows=2, cols=2,
        horizontal_spacing=0.24,
        vertical_spacing=0.28,
        subplot_titles=(
            "Structural Loadings |λ|",
            "Variance Partitioning Profile",
            "Sensor Uniqueness Noise Floor (φ²)",
            "Latent Factor Score Variance"
        )
    )

    # (1,1) Heatmap
    fig.add_trace(go.Heatmap(
        z=np.abs(loadings), x=factor_names, y=sensor_names,
        colorscale='YlOrRd',
        colorbar=dict(x=-0.15, len=0.38, y=0.78, yanchor='middle', xanchor='right'),
        showscale=True
    ), row=1, col=1)

    # (1,2) Stacked bar
    fig.add_trace(go.Bar(
        name='Communality (h²)', y=sensor_names, x=h2 * 100,
        orientation='h', marker_color='#1f77b4', showlegend=True
    ), row=1, col=2)
    fig.add_trace(go.Bar(
        name='Uniqueness (φ²)', y=sensor_names, x=phi2 * 100,
        orientation='h', marker_color='#ff7f0e', showlegend=True
    ), row=1, col=2)
    fig.update_layout(barmode='stack')
    fig.update_xaxes(range=[0, 100], row=1, col=2)

    # (2,1) Uniqueness line
    fig.add_trace(go.Scatter(
        name='Uniqueness φ²', x=sensor_names, y=phi2,
        mode='lines+markers',
        line=dict(color='#d62728', width=2, dash='dashdot'),
        marker=dict(symbol='x', size=10), showlegend=True
    ), row=2, col=1)
    fig.update_xaxes(tickangle=25, row=2, col=1)

    # (2,2) Factor variance bars
    fig.add_trace(go.Bar(
        name='Factor Score Variance', x=factor_names, y=var_factors,
        marker=dict(color='#2ca02c', line=dict(color='black', width=0.5)),
        showlegend=True
    ), row=2, col=2)

    fig.update_layout(
        width=1250, height=750, template='plotly_white',
        legend=dict(orientation='h', x=0.5, y=1.02, xanchor='center'),
        margin=dict(t=150, b=60, l=140, r=80)
    )
    fig.show()

# ─────────────────────────────────────────────
# Run with the assignment synthetic data
# ─────────────────────────────────────────────
np.random.seed(42); n_samples = 2500
f1 = np.random.normal(0,1,n_samples); f2 = np.random.normal(0,1,n_samples)
s1 = 0.85*f1+0.10*f2+np.random.normal(0,0.30,n_samples)
s2 = 0.80*f1+0.15*f2+np.random.normal(0,0.35,n_samples)
s3 = 0.12*f1+0.90*f2+np.random.normal(0,0.25,n_samples)
s4 = 0.02*f1+0.05*f2+np.random.normal(0,1.40,n_samples)
df_asset = pd.DataFrame(np.vstack([s1,s2,s3,s4]).T,
                        columns=['Sensor_1','Sensor_2','Sensor_3','Sensor_4'])

loadings, h2, phi2, F_scores, var_f, sensors = run_fa_engine(df_asset, k=2)
plot_fa_dashboard(loadings, h2, phi2, var_f, sensors)


  Average System Communality : 50.66%
  Average System Uniqueness  : 49.34%
  Sensor                 h²        φ²
  Sensor_1           0.8646    0.1354
  Sensor_2           0.8752    0.1248
  Sensor_3           0.2779    0.7221
  Sensor_4           0.0087    0.9913

  Factor Score Variances: [0.925  0.2841]


# **Subspace Diagnostics & Feature De-correlation**

In structural health monitoring (SHM) and industrial asset diagnostics, multi-sensor grids generate highly continuous, collinear data streams. To establish reliable anomaly detection limits, engineers must project these high-dimensional observation spaces onto lower-dimensional sub-spaces.

This assignment evaluates two fundamental methods for dimensionality reduction using an identical 4-channel sensor array (`Sensor_1` to `Sensor_4`):

1. **Principal Component Analysis (PCA):** A geometric, non-parametric approach aimed at total variance maximization.
2. **Factor Analysis (FA):** A generative, parametric framework aimed at partitioning shared structural variance from unique localized sensor noise.



## **Part A: Develop a Feature Analysis Dashboard**

**UI Layout 1: PCA Optimization Suite ($2\times  3$ Horizontal Layout)**

The first interface must display a 5-panel horizontal grid layout.
* **Panel 1,1 (Heatmap):** Feature Loadings Matrix $|P_{j,r}|$.
* **Panel 1,2 (Bar Chart):** Absolute Eigenvalues ($\lambda_j$) tracking variance magnitude.
* **Panel 1,3 (Combined Bar + Line):** Marginal Explained Variance % per axis (bars) overlaid with Cumulative Captured Variance Curve (dashed line).
* **Panel 2,1 (Bar Chart):** Residual Space (Unexplained Variance %) showing excluded information ($100 - \text{Cumulative Variance \%}$).
* **Panel 2,2 (Scatter Line):** Sample Mean Hotelling's $T^2$ tracking distance vs. Subspace Size $k$.
* **Panel 2,3 (Scatter Line):** Sample Mean $Q$ Statistic (SPE) tracking residual decay vs. Truncation Cutoff $k$.

**UI Layout 2: FA Latent Subspace Suite ($2 \times 2$ Grid Layout)**

The second interface must structure its subplots into a balanced $2 \times 2$ matrix.

* **Panel 1,1 (Heatmap):** Structural Loadings Matrix $|\lambda_{j,r}|$.
* **Panel 1,2 (Stacked Horizontal Bar):** Variance Allocation breakdown showing Shared Communality ($h^2$) versus Channel Uniqueness ($\varphi^2$) totaling $100\%$ per sensor.
* **Panel 2,1 (Scatter Line):** Sensor Uniqueness Noise Floor Profile ($\varphi^2$) across monitored channels to instantly isolate instrument faults.
* **Panel 2,2 (Bar Chart):** Latent Factor Scores Empirical Variance showing the post-rotation power of the extracted physical dimensions.


---
### Answer: Part A
---


In [4]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import warnings
warnings.filterwarnings('ignore')

# ── Synthetic asset data ───────────────────────────────────
np.random.seed(42); n_samples = 2500
f1 = np.random.normal(0,1,n_samples); f2 = np.random.normal(0,1,n_samples)
s1 = 0.85*f1+0.10*f2+np.random.normal(0,0.30,n_samples)
s2 = 0.80*f1+0.15*f2+np.random.normal(0,0.35,n_samples)
s3 = 0.12*f1+0.90*f2+np.random.normal(0,0.25,n_samples)
s4 = 0.02*f1+0.05*f2+np.random.normal(0,1.40,n_samples)
df_asset = pd.DataFrame(np.vstack([s1,s2,s3,s4]).T,
                        columns=['Sensor_1','Sensor_2','Sensor_3','Sensor_4'])
sensor_names = df_asset.columns.tolist()

# ── Standardize ────────────────────────────────────────────
X = df_asset.values.astype(float)
Z = (X - X.mean(0)) / X.std(0, ddof=1)
m = Z.shape[1]

# ── PCA via eigh ───────────────────────────────────────────
S = np.cov(Z, rowvar=False)
evals, evecs = np.linalg.eigh(S)
idx = np.argsort(evals)[::-1]; evals = evals[idx]; evecs = evecs[:, idx]
scores_all = Z @ evecs
total_var  = evals.sum()
cum_var    = np.cumsum(evals) / total_var * 100
marginal   = evals / total_var * 100
residual   = 100 - cum_var
pc_names   = [f'PC {j+1}' for j in range(m)]
k_vals     = [1, 2, 3]
mean_T2    = [np.sum((scores_all[:,:k]**2)/evals[:k],axis=1).mean() for k in k_vals]
mean_Q     = [np.sum(scores_all[:,k:]**2,axis=1).mean() for k in k_vals]

# ── FA (k=2, Varimax) ──────────────────────────────────────
fa = FactorAnalysis(n_components=2, rotation='varimax', random_state=42)
fa.fit(Z)
fa_loadings = fa.components_.T
h2   = np.sum(fa_loadings**2, axis=1)
phi2 = 1 - h2
R    = np.corrcoef(Z, rowvar=False)
F_sc = Z @ np.linalg.inv(R) @ fa_loadings
var_f = F_sc.var(axis=0, ddof=1)
factor_names = ['Factor 1', 'Factor 2']

# ════════════════════════════════════════════
# UI Layout 1: PCA Dashboard (2×3)
# ════════════════════════════════════════════
fig1 = make_subplots(
    rows=2, cols=3,
    horizontal_spacing=0.16, vertical_spacing=0.30,
    subplot_titles=(
        'Feature Loadings |P|',
        'Eigenvalue Magnitudes (λ)',
        'Marginal & Cumulative Variance %',
        'Residual Unexplained Variance %',
        'Mean Hotelling T² vs k',
        'Mean Q Statistic (SPE) vs k'
    )
)

# (1,1) PCA loadings heatmap
fig1.add_trace(go.Heatmap(
    z=np.abs(evecs), x=pc_names, y=sensor_names,
    colorscale='Blues', showscale=True,
    colorbar=dict(len=0.4, y=0.78)
), row=1, col=1)

# (1,2) Eigenvalues
fig1.add_trace(go.Bar(
    x=pc_names, y=evals, marker_color='#1f77b4',
    marker_line=dict(color='black', width=0.5), name='Eigenvalue'
), row=1, col=2)

# (1,3) Marginal bars + cumulative line
fig1.add_trace(go.Bar(
    x=pc_names, y=marginal, marker_color='#aec7e8', name='Marginal Var %'
), row=1, col=3)
fig1.add_trace(go.Scatter(
    x=pc_names, y=cum_var, mode='lines+markers',
    line=dict(dash='dash', color='#d62728', width=2),
    marker=dict(size=8), name='Cumulative Var %'
), row=1, col=3)

# (2,1) Residual variance
fig1.add_trace(go.Bar(
    x=pc_names, y=residual, marker_color='#ff7f0e',
    marker_line=dict(color='black', width=0.5), name='Residual Var %'
), row=2, col=1)

# (2,2) Mean T²
fig1.add_trace(go.Scatter(
    x=k_vals, y=mean_T2, mode='lines+markers',
    line=dict(color='#2ca02c', width=2),
    marker=dict(size=10, symbol='circle'), name='Mean T²'
), row=2, col=2)

# (2,3) Mean Q
fig1.add_trace(go.Scatter(
    x=k_vals, y=mean_Q, mode='lines+markers',
    line=dict(color='#9467bd', width=2),
    marker=dict(size=10, symbol='diamond'), name='Mean Q (SPE)'
), row=2, col=3)

fig1.update_layout(
    title_text='PCA Optimization Suite', width=1300, height=750,
    template='plotly_white', showlegend=False,
    margin=dict(t=120, b=60, l=80, r=60)
)
fig1.show()

# ════════════════════════════════════════════
# UI Layout 2: FA Dashboard (2×2)
# ════════════════════════════════════════════
fig2 = make_subplots(
    rows=2, cols=2,
    horizontal_spacing=0.24, vertical_spacing=0.28,
    subplot_titles=(
        'Structural Loadings |λ|',
        'Variance Partitioning Profile',
        'Sensor Uniqueness Noise Floor (φ²)',
        'Latent Factor Score Variance'
    )
)

fig2.add_trace(go.Heatmap(
    z=np.abs(fa_loadings), x=factor_names, y=sensor_names,
    colorscale='YlOrRd',
    colorbar=dict(x=-0.15, len=0.38, y=0.78, yanchor='middle', xanchor='right')
), row=1, col=1)

fig2.add_trace(go.Bar(
    name='Communality (h²)', y=sensor_names, x=h2*100,
    orientation='h', marker_color='#1f77b4'
), row=1, col=2)
fig2.add_trace(go.Bar(
    name='Uniqueness (φ²)', y=sensor_names, x=phi2*100,
    orientation='h', marker_color='#ff7f0e'
), row=1, col=2)
fig2.update_layout(barmode='stack')
fig2.update_xaxes(range=[0,100], row=1, col=2)

fig2.add_trace(go.Scatter(
    name='Uniqueness φ²', x=sensor_names, y=phi2, mode='lines+markers',
    line=dict(color='#d62728', width=2, dash='dashdot'),
    marker=dict(symbol='x', size=10)
), row=2, col=1)
fig2.update_xaxes(tickangle=25, row=2, col=1)

fig2.add_trace(go.Bar(
    name='Factor Score Variance', x=factor_names, y=var_f,
    marker=dict(color='#2ca02c', line=dict(color='black', width=0.5))
), row=2, col=2)

fig2.update_layout(
    title_text='FA Latent Subspace Suite', width=1250, height=750,
    template='plotly_white',
    legend=dict(orientation='h', x=0.5, y=1.02, xanchor='center'),
    margin=dict(t=150, b=60, l=140, r=80)
)
fig2.show()


## **Part B: Analysis**

### Generate the following synthetic data and answer the questions below

The synthetic multi-sensor asset data must be generated using
```python
import numpy as np
import pandas as pd

# Set execution seed for perfect reproducible properties
np.random.seed(42)
n_samples = 2500

# 1. Generate underlying, independent true physical driving forces (Latent Factors)
f1 = np.random.normal(0, 1, n_samples)  # Primary structural mode/load
f2 = np.random.normal(0, 1, n_samples)  # Secondary operational mode

# 2. Construct the Observable Multi-Sensor Array with targeted noise profiles
# Sensors 1 and 2 couple strongly to the primary process (f1)
s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)

# Sensor 3 couples strongly to the secondary process (f2)
s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)

# Sensor 4 suffers an extreme electrical/instrumentation fault (massive localized white noise)
s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

# 3. Compile into the final evaluation DataFrame
df_asset = pd.DataFrame(
    data=np.vstack([s1, s2, s3, s4]).T,
    columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
)

```

**Empirical Covariance and Correlation Properties**

* **Sensors 1, 2, and 3:** High cross-correlation due to strong coupling with the common underlying physical structures ($f_1$ and $f_2$).
* **Sensor 4:** Highly un-correlated with all other channels, but exhibits an exceptionally large variance magnitude ($\sigma^2 \approx 2.0$) due to its high noise contamination scale.


### Question 1: The Total Variance Illusion (PCA Subplots 1 & 2 vs. FA Subplots 1 & 2)

* **Context:** In your PCA dashboard, `PC 1` captures roughly $45\%$ of the total system variance, and `PC 2` captures over $30\%$, indicating a highly structured, multi-component system. However, looking at the Factor Analysis dashboard, the **Variance Partitioning Stacked Bar** reveals that `Sensor_4` has a **Uniqueness value ($\varphi^2$) exceeding $98\%$**.
* **Core Problem:** 1. What does a Uniqueness value close to $100\%$ tell you about the physical nature of the data coming from `Sensor_4`?
2. Because PCA maximizes total global variance without distinguishing between types of variance, how does the massive localized noise of `Sensor_4` "trick" the PCA engine into inflating its top eigenvalues?
3. Why is this a major risk if an engineer relies *only* on PCA to build an automated plant anomaly detection framework?



---
###Answer: Question 1
---


**1. Uniqueness ≈ 99% for Sensor 4:**

A uniqueness value close to 100% means that virtually **none** of Sensor 4's variance is shared with the other sensors through the common latent factors. Its signal is almost entirely an isolated, localised noise process — in this case a severe electrical fault modelled as high-amplitude white noise ($\sigma \approx 1.40$). The sensor is statistically decoupled from the structural dynamics and provides no useful information about the physical system.

**2. How Sensor 4's noise "tricks" PCA:**

PCA maximises total explained variance without distinguishing structural signal from measurement noise. Because Sensor 4 has raw variance $\approx 2.0$ — far larger than the other three channels ($\approx 0.8$–$1.1$) — it dominates the off-diagonal structure of the sample covariance matrix. The first eigenvector is pulled toward directions that explain a large fraction of Sensor 4's noise, inflating the apparent eigenvalue of PC1 and distorting the rotation of the entire principal frame. The top PCs absorb this noise energy and present it as "signal."

**3. Why PCA-only monitoring is a major risk:**

If an engineer builds an anomaly detector on $T^2$ thresholds calibrated from a PCA model contaminated by Sensor 4's noise:
- The monitoring ellipsoid is calibrated to a noisy, distorted coordinate system.
- A real structural anomaly shifting the mean of Sensors 1–3 may produce a smaller $T^2$ increase than Sensor 4's background noise, causing the fault to go undetected (inflated Type II error rate).
- A purely instrumental degradation of Sensor 4 can trigger false alarms because the model encoded that noise as "normal" structure.
FA separates shared structural variance from sensor-specific noise, making diagnostic boundaries far more meaningful.


### Question 2: Decoupling Structural Loading via Rotation (FA Subplot 1 vs. PCA Eigenvectors)

* **Context:** Look at the **Structural Loadings Matrix** heatmap in the top-left of the FA dashboard. You can see that `Sensor_1` and `Sensor_2` load heavily on `Factor 1`, while `Sensor_3` loads exclusively on `Factor 2`.
* **Core Problem:**
1. Traditional PCA forces its principal axes to follow a strict mathematical hierarchy ($\lambda_1 > \lambda_2 > \dots$) where every component tries to sweep up as much mixed variance as possible across all sensors. Explain how the **Varimax rotation** in Factor Analysis relaxes this rule to create "simple structure" (clean, isolated groupings).
2. From a plant operator's standpoint, why is the rotated FA heatmap much easier to troubleshoot during a structural failure than a raw, unrotated PCA eigenvector matrix?





---
### Answer: Question 2
---


**1. How Varimax creates "simple structure":**

PCA eigenvectors are constrained to the directions of sequentially maximum variance, producing dense vectors where every sensor has a non-trivial loading on every component. Varimax applies an orthogonal rotation matrix $\mathbf{T}$ to the initial loading estimate, maximising the variance of the squared loadings within each factor column:
$$V = \frac{1}{m}\sum_{r=1}^k\left[\sum_j \lambda_{jr}^4 - \frac{1}{m}\left(\sum_j \lambda_{jr}^2\right)^2\right]$$
This drives each factor toward large loadings on a small cluster of sensors and near-zero loadings on the rest. Unlike PCA, Varimax does not enforce a variance-ordering hierarchy; it trades interpretability for optimality-of-allocation, yielding Factor 1 ↔ Sensors 1 & 2, Factor 2 ↔ Sensor 3.

**2. Why the FA heatmap is easier to use during a real failure:**

In a raw PCA eigenvector matrix every sensor has a moderate mixed contribution to every PC, requiring mental linear algebra to isolate the faulty subsystem. In the Varimax-rotated FA heatmap each factor maps directly to an identifiable physical mode: a plant operator can immediately read "Factor 1 anomaly → primary load path (Sensors 1 & 2); Factor 2 anomaly → secondary mode (Sensor 3)." Troubleshooting becomes a simple table look-up rather than a matrix computation.


### Question 3: Determining Subspace Truncation ($k$) using $T^2$ and $Q$ Profiles

* **Context:** Look at the right side of the PCA dashboard, showing **Mean Hotelling's $T^2$** and **Mean $Q$ Statistic (SPE)** plotted against the truncation cutoff $k$.
* **Core Problem:**
1. Describe the exact behavior of the Mean $Q$ Statistic curve as the subspace cutoff transitions from $k=1$ to $k=2$, and then from $k=2$ to $k=3$.
2. Why does a sharp drop followed by a distinct "flat elbow" in the $Q$ statistic identify **$k=2$** as the true hidden physical dimensionality of this asset?
3. If you incorrectly choose a truncation cutoff of $k=3$ for your monitoring system, what type of data are you forcing into your "clean" principal subspace?

---
### Answer: Question 3
---


**1. Behaviour of Mean Q as k increases:**

| k | Mean T² | Mean Q |
|---|---------|--------|
| 1 | ≈ 1.00 | ≈ 2.02 |
| 2 | ≈ 2.00 | ≈ 1.02 |
| 3 | ≈ 3.00 | ≈ 0.14 |

From $k=1$ to $k=2$, Mean $Q$ drops sharply (≈ 50%) because PC2 captures a second genuine structural mode (the secondary factor $f_2$). From $k=2$ to $k=3$, Mean $Q$ falls again but by a much smaller amount and enters a noise floor — this final reduction reflects mostly the white-noise energy of Sensor 4.

**2. Why the flat elbow at $k=2$ identifies the true dimensionality:**

The sharp drop from $k=1$ to $k=2$ indicates that PC2 accounts for a large, real residual signal. The plateau (flat elbow) after $k=2$ indicates that PC3 and beyond are absorbing measurement noise, not structured variance. The $Q$ statistic "bottoms out" into a noise floor at $k=2$, signalling that the two retained PCs already explain all coherent physical information. This is the scree/elbow criterion applied to the residual domain.

**3. Risk of choosing $k=3$:**

With $k=3$ you force the **noise-dominated direction** of Sensor 4 (driven by its white noise) into the "clean" principal subspace. The $T^2$ control limit must now account for this noise dimension. As a result, real structural anomalies produce smaller relative $T^2$ deviations and are more easily missed, and the healthy operation ellipsoid is inflated in the Sensor 4 noise direction, degrading sensitivity to genuine structural changes.


### Question 4: Operational Trade-offs in System Health Monitoring

* **Context:** Imagine you are deploying a real-time predictive maintenance pipeline for a fleet of identical industrial assets based on these exact models.
* **Core Problem:** Compare the operational trade-offs of these two approaches:
1. **The PCA Strategy:** Tracking the main principal subspace using Hotelling's $T^2$ alongside a residual noise floor monitor using the $Q$ statistic.
2. **The FA Strategy:** Monitoring the rotated Latent Factor Scores directly.


* Which strategy is more robust against a single sensor losing calibration or experiencing an electrical short? Justify your choice using the **Sensor Uniqueness Noise Floor ($\varphi^2$)** metrics.

---
### Answer: Question 4
---


**PCA Strategy (T² + Q monitoring):**

*Strengths:* Computationally simple; mathematically optimal for total variance capture; the $Q$ statistic is sensitive to any deviation from the nominal subspace, including new fault modes not seen during training.

*Weaknesses:* Does not separate structural variance from sensor-specific noise. A single sensor losing calibration inflates its raw variance, which distorts the entire eigenvector frame. The model must be retrained, and during the degradation period the $T^2$ and $Q$ thresholds are miscalibrated. Sensor 4's high uniqueness ($\varphi^2 \approx 99\%$) would not be flagged as an instrument fault — it would silently corrupt the model geometry.

**FA Strategy (Latent Factor Score monitoring):**

*Strengths:* Explicitly decomposes each sensor's variance into shared ($h^2$) and isolated ($\varphi^2$) components. A sensor fault manifests as a sudden jump in that sensor's uniqueness $\varphi^2$ without contaminating the shared factor scores. Monitoring factor scores is equivalent to monitoring the physically interpretable structural modes.

*Weaknesses:* Assumes a specific generative model (linear, Gaussian); more sensitive to model mis-specification; factor extraction is not unique without rotation constraints.

**Conclusion — the FA strategy is more robust against a single sensor fault.**

The Sensor Uniqueness Noise Floor profile ($\varphi^2$) provides a direct, sensor-level diagnostic: if Sensor 4 loses calibration, its $\varphi^2$ jumps toward 1.0, immediately flagging the instrument fault without affecting the Factor 1 and Factor 2 score statistics. The healthy structural modes continue to be monitored accurately through the rotated factor scores of the remaining high-communality sensors (Sensors 1–3 with $\varphi^2 \approx 0.13$, $0.12$, and $0.72$ respectively).

In the PCA strategy, the same fault forces a full model rebuild, as the corrupted sensor's variance contaminates the eigenvectors of all PCs. Until retraining, both $T^2$ and $Q$ operate with a distorted coordinate frame, making it impossible to reliably separate instrument drift from genuine structural damage.
